### PR TITLE
Add possibility to opt out of `stdio.h` and `stdarg.h`

### PR DESCRIPTION
**Pull request by albinahlback**
Almost a 10% reduction in compile time when opting out of them during the build (and that the build should not affect the library).

Remove all inclusions of `limits.h` as GMP already includes it.

Only include `sys/param.h` if user uses Cygwin.

Try to not include `string.h` in source files.

Add TEST_CFLAGS in configure script in order to use different compiler flags for the tests versus build.

### CODE REVIEW

**Comment by fredrik-johansson** [comment]
BSD failed?

**Comment by albinahlback** [comment]
Yeah, I'm not sure why exactly. I will restart it once more.

**Comment by fredrik-johansson** [comment]
It may be one of those tests that is failing with test multiplier < 1, but strange that it passed before (the RNG should be deterministic).

File: src/fmpz_mpoly/mpolyd.c [hunk 1]
```diff
@@ -9,6 +9,12 @@
     (at your option) any later version.  See <https://www.gnu.org/licenses/>.
 */
 
+#ifdef __GNUC__
+# define printf __builtin_printf
+#else
+# include <stdio.h>
+#endif
+
```
Comment (fredrik-johansson) [inline src/fmpz_mpoly/mpolyd.c:17]: Why not change the printf to flint_printf?

File: src/fmpz_poly/fread_pretty.c [hunk 2]
```diff
@@ -9,10 +9,16 @@
...
+#include <stdio.h>
 #include <ctype.h>
 
-#include "flint.h"
+#ifdef __GNUC__
+# define strcmp __builtin_strcmp
+# define strcpy __builtin_strcpy
+#else
+# include <string.h>
+#endif
```
Comment (fredrik-johansson) [inline src/fmpz_poly/fread_pretty.c:20]: Is there any real benefit to avoiding standard includes in single files?

File: src/fmpz_mpoly/mpolyd.c (hunk 1 above)
Comment (albinahlback) [reply src/fmpz_mpoly/mpolyd.c:17]: As the PR is pretty large, I want to keep a theme and try to leave the source code alone (I believe I changed some code in some header from printf to flint_printf in order to get rid of `stdio.h`).

I believe that all internals should use `printf`, but here I just didn't want to touch the code in the function.

File: src/fmpz_poly/fread_pretty.c (hunk 2 above)
Comment (albinahlback) [reply src/fmpz_poly/fread_pretty.c:20]: No, not really. Perhaps a second faster build if done throughout the repository. Although unnecessary inclusions should be removed, doing this for the necessary ones won't really save you time.

Comment (albinahlback) [reply src/fmpz_poly/fread_pretty.c:20]: I thought it was going to make a difference when I started out. Do you want me to remove them?

**Comment by albinahlback** [comment]
It counts probable primes, where its acceptable limit is 110 but it gets 157.

**Comment by fingolfin** [comment]
Great!

May I ask why this is opt-out instead of opt-in? I am guessing to make the transition easier? I think it would be good to make the reasons explicit.

Does this result in identical binaries? I hope so -- then a single libflint can be installed and it's up packages using FLINT to decide whether they want to enable or disable these header defines.

Here is an argument for making it opt-in: for many projects all will work, and those which really need it can easily opt-in -- or better, just explicit uses `#include <stdio.h>` wherever they need this.

(Mind you, I am not complaining anyway, just some food for thought. Thank you for working on this!)

**Comment by albinahlback** [comment]
> Great!
> 
> May I ask why this is opt-out instead of opt-in? I am guessing to make the transition easier? I think it would be good to make the reasons explicit.

To be backwards compatible (although that's not very important for me, Fredrik is the one who's responsible for FLINT ;-)).


> Does this result in identical binaries? I hope so -- then a single libflint can be installed and it's up packages using FLINT to decide whether they want to enable or disable these header defines.

That's the idea. The only problem is that we have too many inline functions that require the use of `stdio.h`, otherwise they will default to the non-inlined version. As not all source files include `stdio.h` (this would have to be changed in order to merge this PR), this would result in different binaries.


> Here is an argument for making it opt-in: for many projects all will work, and those which really need it can easily opt-in -- or better, just explicit uses `#include <stdio.h>` wherever they need this.

Those that could gain a lot from this is developers of larger projects, while this can be very confusing for beginners. Developers will see a speed-up, but beginners will not know why you need to include headers in order to utilize functions, why order of inclusions matter and why their code broke.

**Comment by fredrik-johansson** [comment]
> That's the idea. The only problem is that we have too many inline functions that require the use of stdio.h, otherwise they will default to the non-inlined version. As not all source files include stdio.h (this would have to be changed in order to merge this PR), this would result in different binaries.

Inlining isn't the issue. None of the IO functions need to have inlined version. The only reason we have inlines like

    FMPQ_INLINE int fmpq_print(const fmpq_t x)
    {
        return fmpq_fprint(stdout, x);
    }

is that this was less effort than creating a separate .c file for this function, but all the fmpq print functions could go in ``fmpq/print.c``, for example.

I think the only use for globally including ``<stdio.h>`` is to have the ``FILE`` type for ``fprint`` functions and such. But perhaps the file IO functions are rare enough that it's acceptable to require users to include ``<stdio.h>`` first?

We do need ``FILE`` in some typedefs, but perhaps they could be replaced with ``void *`` together with appropriate use of casts.

**Comment by albinahlback** [comment]
> Inlining isn't the issue. None of the IO functions need to have inlined version.

In regards to producing indentical binaries, this is the only issue with this PR.

File: src/fmpq.h [hunk 3]
```diff
@@ -202,6 +205,10 @@ FMPQ_INLINE int fmpq_print(const fmpq_t x)
 {
     return fmpq_fprint(stdout, x);
 }
+#else
+int _fmpq_print(const fmpz_t num, const fmpz_t den);
+int fmpq_print(const fmpq_t x);
```
Comment (fingolfin) [inline src/fmpq.h:210]: Maybe just de-inline `fmpq_print` and `_fmpq_print` and their relatives everywhere? Then that's one less thing that is different based on `FLINT_HAVE_FILE`

File: src/flint.h [hunk 4]
```diff
@@ -438,14 +451,20 @@ mp_limb_t FLINT_BIT_COUNT(mp_limb_t x)
...
 FLINT_DLL int flint_scanf(const char * str, ...); /* flint version of scanf */
-FLINT_DLL int flint_fscanf(FILE * f, const char * str, ...); /* flint version of fscanf */
 FLINT_DLL int flint_sscanf(const char * s, const char * str, ...); /* flint version of sscanf */
 
+#ifdef FLINT_HAVE_VA_LIST
+FLINT_DLL int flint_vprintf(const char * str, va_list ap); /* va_list version of flint_printf */
+#endif
+
+#ifdef FLINT_HAVE_FILE
+FLINT_DLL int flint_fprintf(FILE * f, const char * str, ...); /* flint version of fprintf */
```
Comment (fingolfin) [inline src/flint.h:464]: Have you considered just doing `typedef void * FLINT_FILE` somewhere and using that in those headers?

File: src/fmpq.h (hunk 3 above)
Comment (albinahlback) [reply src/fmpq.h:210]: Yes, I think both me and Fredrik believe they should not be inlined. But that should be a different PR

File: src/flint.h (hunk 4 above)
Comment (albinahlback) [reply src/flint.h:464]: Sorry for my late answer. So I do not like replacing `FILE` just like that everywhere. If you have a filestream you want to use, you have obviously included `stdio.h`. However, for other types where you typically initialize a struct with a filestream, such as stream objects in `gr.h`, this is actually a great idea. I am first going to do a PR where I deinline printer functions, and then we can argue whether we should include `stdio.h` by default or not ;-)

**Comment by albinahlback** [comment]
Close in favor of #1283

**Pull request closed without merging**